### PR TITLE
Implementation of LambdaSystem wrapper for Update functions for syste…

### PR DESCRIPTION
**Pull request by JDuchniewicz**
…ms, Implementation of ISystem

### CODE REVIEW

File: PolyEngine/Engine/Src/ECS/LambdaSystem.cpp [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#include "EnginePCH.hpp"
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.cpp]: Let's use <> in every include from now on.

File: PolyEngine/Engine/Src/ECS/LambdaSystem.hpp [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#pragma once
+
+#include <Defines.hpp>
+#include "ECS/ISystem.hpp"
+
+namespace Poly
+{
+	class Scene;
+	typedef std::function<void(Scene*)> PhaseUpdateFunction;
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.hpp]: change to using

File: PolyEngine/Engine/Src/ECS/LambdaSystem.hpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+	class LambdaSystem : public ISystem
+	{
+	public:
+		LambdaSystem(PhaseUpdateFunction func) { Payload = func; }
+		void OnUpdate(Scene* world) { Payload(world); }
+	private:
+		PhaseUpdateFunction Payload;
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.hpp]: `Payload` sounds strange here. Rename to `UpdateFunctionObject`

File: PolyEngine/Engine/Src/ECS/LambdaSystem.hpp [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Poly
+{
+	class Scene;
+	typedef std::function<void(Scene*)> PhaseUpdateFunction;
+
+	class LambdaSystem : public ISystem
+	{
+	public:
+		LambdaSystem(PhaseUpdateFunction func) { Payload = func; }
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.hpp]: `Payload = std::move(func)`

File: PolyEngine/Engine/Src/Engine.hpp [hunk 5]
```diff
@@ -173,7 +178,7 @@ namespace Poly
 		OpenALDevice AudioDevice;
 		InputQueue InputEventsQueue;
 
-		Dynarray<PhaseUpdateFunction> GameUpdatePhases[static_cast<int>(eUpdatePhaseOrder::_COUNT)];
+		Dynarray<ISystem*> GameUpdatePhases[static_cast<int>(eUpdatePhaseOrder::_COUNT)]; //change to array of uniqueptrs
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: This should hold unique ptr to ISystem.

File: PolyEngine/Engine/Src/Engine.hpp [hunk 6]
```diff
@@ -155,16 +158,18 @@ namespace Poly
...
 
 		/// Registers a PhaseUpdateFunction to be executed in the update.
 		/// part of a single frame in the same order as they were passed in.
 		/// @param phaseFunction - void function(Scene*)
 		/// @param order - enum eUpdatePhaseOrder value
 		/// @see eUpdatePhaseOrder
-		void RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order);
+		void RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order); //add for Isystems
+
+		void RegisterUpdatePhase(ISystem* system, eUpdatePhaseOrder order);
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: Change this one to RegisterSystem and make it accept unique ptr

File: PolyEngine/Engine/Src/Engine.hpp [hunk 7]
```diff
@@ -155,16 +158,18 @@ namespace Poly
...
+				update->OnUpdate(GetActiveScene());
 		}
 
 		/// Registers a PhaseUpdateFunction to be executed in the update.
 		/// part of a single frame in the same order as they were passed in.
 		/// @param phaseFunction - void function(Scene*)
 		/// @param order - enum eUpdatePhaseOrder value
 		/// @see eUpdatePhaseOrder
-		void RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order);
+		void RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order); //add for Isystems
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: add [[deprecated]] attribute

File: PolyEngine/Engine/Src/Engine.cpp [hunk 8]
```diff
@@ -99,10 +100,20 @@ Engine::~Engine()
 void Engine::RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order)
 {
 	HEAVY_ASSERTE(order != eUpdatePhaseOrder::_COUNT, "_COUNT enum value passed to RegisterUpdatePhase(), which is an invalid value");
-	Dynarray<PhaseUpdateFunction>& UpdatePhases = GameUpdatePhases[static_cast<int>(order)];
-	UpdatePhases.PushBack(phaseFunction);
+	Dynarray<ISystem*>& UpdatePhases = GameUpdatePhases[static_cast<int>(order)];
+	std::unique_ptr<ISystem> system = std::make_unique<ISystem>(LambdaSystem(phaseFunction));
+	UpdatePhases.PushBack(system.get());
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.cpp]: This is freed upon leaving this scope.

File: PolyEngine/Engine/Src/Engine.hpp (hunk 7 above)
Comment (JDuchniewicz) [reply PolyEngine/Engine/Src/Engine.hpp]: marking this deprecated causes compiler errors (warning treated as errors), as this method is still being used

File: PolyEngine/Engine/Src/Engine.hpp [hunk 9]
```diff
@@ -48,7 +48,8 @@ namespace Poly
 		/// <param name="device">Pointer to IRenderingDevice instance.</param>
 		void Init(std::unique_ptr<IGame> game, std::unique_ptr<IRenderingDevice> device);
 
-		/// <summary>Registers a PhaseUpdateFunction to be executed in the update.</summary>
+		/// <summary>DEPRECATED!!! Do not use!
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: Please use `@deprecated`

https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmddeprecated

File: PolyEngine/Engine/Src/Engine.hpp [hunk 10]
```diff
@@ -164,13 +165,19 @@ namespace Poly
 				update->OnUpdate(GetActiveScene());
 		}
 
+		/// DEPRECATED!!! Do not use!
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: Please use `@deprecated`

File: PolyEngine/Engine/Src/ECS/ISystem.hpp [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <Defines.hpp>
+
+namespace Poly
+{
+	class Scene;
+	class ISystem
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/ISystem.hpp]: Derive from `BaseObject<>`

File: PolyEngine/Engine/Src/ECS/LambdaSystem.hpp [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+namespace Poly
+{
+	class Scene;
+	using PhaseUpdateFunction = std::function<void(Scene*)>;
+
+	class LambdaSystem : public ISystem
+	{
+	public:
+		LambdaSystem(PhaseUpdateFunction func) { UpdateFunctionObject = std::move(func); }
+		void OnUpdate(Scene* scene) { UpdateFunctionObject(scene); }
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.hpp]: Add `override` keyword

File: PolyEngine/Engine/Src/Engine.hpp [hunk 13]
```diff
@@ -155,25 +161,33 @@ namespace Poly
...
+
 		std::unique_ptr<Scene> ActiveScene;
 		Scene* SerializedScene = nullptr;
 		std::unique_ptr<IGame> Game;
 		std::unique_ptr<IRenderingDevice> RenderingDevice;
 		OpenALDevice AudioDevice;
 		InputQueue InputEventsQueue;
 
-		Dynarray<PhaseUpdateFunction> GameUpdatePhases[static_cast<int>(eUpdatePhaseOrder::_COUNT)];
+		Dynarray<std::unique_ptr<ISystem>> GameUpdatePhases[static_cast<int>(eUpdatePhaseOrder::_COUNT)]; //change to array of uniqueptrs
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.hpp]: Let's refactor this array to `EnumArray`

File: PolyEngine/Engine/Src/ECS/LambdaSystem.hpp [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+namespace Poly
+{
+	class Scene;
+	using PhaseUpdateFunction = std::function<void(Scene*)>;
+
+	class LambdaSystem : public ISystem
+	{
+	public:
+		LambdaSystem(PhaseUpdateFunction func) { UpdateFunctionObject = std::move(func); }
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/ECS/LambdaSystem.hpp]: Make constructor `explicit`

File: PolyEngine/Engine/Src/Engine.cpp [hunk 15]
```diff
@@ -99,10 +100,19 @@ Engine::~Engine()
 void Engine::RegisterUpdatePhase(const PhaseUpdateFunction& phaseFunction, eUpdatePhaseOrder order)
 {
 	HEAVY_ASSERTE(order != eUpdatePhaseOrder::_COUNT, "_COUNT enum value passed to RegisterUpdatePhase(), which is an invalid value");
-	Dynarray<PhaseUpdateFunction>& UpdatePhases = GameUpdatePhases[static_cast<int>(order)];
-	UpdatePhases.PushBack(phaseFunction);
+	Dynarray<std::unique_ptr<ISystem>>& UpdatePhases = GameUpdatePhases[static_cast<int>(order)];
+	UpdatePhases.PushBack(std::make_unique<LambdaSystem>(LambdaSystem(phaseFunction)));
```
Comment (MuniuDev) [inline PolyEngine/Engine/Src/Engine.cpp]: No need to call constructor, make_unique already does.

**Pull request merged**